enhancements/20240506-cpu-pinning.md

```diff
@@ -0,0 +1,219 @@
+that we want to change in the `.stringdata.configs` field. It should be a json format like the following example.
+In the following example, we want to change node-0 to none policy, node-1 to static policy, node-2 to none policy.
+```json
+{
+  "configs": [
+    {
+      "name": "node-0",
+      "policy": "none"
+    },
+    {
+      "name": "node-1",
+      "policy": "static"
+    },
+    {
+      "name": "node-2",
+      "policy": "none"
+    }
+  ]
+}
+```
+- name: node name
+- policy: cpu manager policy name. should be either `static` or `none`.
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cpu-manager
+  namespace: cattle-system
+type: Opaque
+stringData:
+  configs: '{"configs": [{"name": "node-0", "policy": "none"}, {"name": "node-1", "policy": "static", {"name": "node-2", "policy": "none"}]}'
+---
+apiVersion: upgrade.cattle.io/v1
+kind: Plan
+metadata:
+  name: check-rke2-server
+  namespace: cattle-system
+spec:
+  concurrency: 1
+  nodeSelector:
+    matchLabels:
+      harvesterhci.io/managed: "true"
+  serviceAccountName: system-upgrade-controller
+  version: v1.1.0
+  secrets:
+    - name: cpu-manager
+      path: /cpu-manager
+  upgrade:
+    image: registry.suse.com/bci/bci-base:15.5
+    command: ["/bin/sh", "-c"]
+    args:
+      # see the following description
+```
+`.spec.upgrade.args` section should do the following things:
+1. check if the current node id is in `/cpu-manager/configs`, if yes, go to next step, if not, do nothing and return.
+2. check if `--cpu-manager-policy` is not the same as the node policy in /cpu-manager/configs json file, if yes, go to next step, if not, do nothing and return.
+3. modify `/etc/rancher/rke2/config.yaml.d/90-harvester-server.yaml` and change the `--cpu-manager-policy`
+4. remove `/var/lib/kubelet/cpu_manager_state`
+5. restart rke2-server or rke2-agent (where the kubelet belongs to)
+
+Once the above implementations are completed, we can activate VM CPU pinning by including `dedicatedCpuPlacement=true`
+in `.spec.template.cpu`. Additionally, ensuring that CPU limits and requests are identical and integers (i.e., Guaranteed QoS).
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  template:
+    spec:
+      domain:
+        cpu:
+          cores: 2
+          sockets: 1
+          threads: 1
+          dedicatedCpuPlacement: true
+        resources:
+          limits:
+            cpu: 2
+          requests:
+            cpu: 2
+[...]
+```
+
+#### Web UI
+Regarding the Web UI, we should display the number of CPUs already pinned on each node,
+as well as the remaining number of CPUs available for allocation on each node.
+
+### Test plan
+TBD
+
+## Discussion
+- How much CPU resource allocation should be reserved? We have to set cpu resources in either `--kube-reserved` or `--system-reserved`
```

Comment:
> Note:System services such as the container runtime and the kubelet itself can continue to run on these exclusive CPUs. The exclusivity only extends to other pods.

Thanks for the comment, @w13915984028! 😃 
Indeed, that's why I'm struggling with determining how much CPU resources should we reserved. If system services still use CPUs allocated to VMs with CPU pinning, the concept of “exclusive CPUs” becomes meaningless.

> reserve: max (4, 20% of CPU cores); others for pin-able usage. ?

Sounds good, while I'm wondering when should we setup the CPU resource reservation ?
1. Whenever we want to setup static policy in each harvester node.
2. During the installation of harvester.

I would lean to option2 since it is simpler. To me, CPU resource reservation is more of a setting that harvester should be aware of from the beginning, not just for this feature (CPU Pinning).

Reply:
Sorry for the late response, now we use formula in https://cloud.google.com/kubernetes-engine/docs/concepts/plan-node-sizes  to calculate `kube-reserved` and `system-reserved` cpu resource.